### PR TITLE
fix HashMethodOneNumber with const column

### LINKED ISSUE #39818: Wrong result of intersect
```sql
select 1 from numbers(10) intersect select 1 from numbers(10)

SELECT 1
FROM numbers(10)
INTERSECT
SELECT 1
FROM numbers(10)

Query id: 9f2d11c5-269c-41dc-9b4c-8f1423aafcc7

┌─1─┐
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
│ 1 │
└───┘

8 rows in set. Elapsed: 0.002 sec. 
```

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
fix HashMethodOneNumber get wrong key value when column is const

Close #39818 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by Avogar** [comment]
Thanks! Can you add a test for it?

**Comment by Avogar** [comment]
Also worth fixing all HashMethod*, not only HashMethodOneNumber, for example such query leads to segfault:
```
select '1' from numbers(10) intersect select '1' from numbers(10)
```
If you have time, you can fix it in this PR, if not, I will create a separate PR.

**Comment by canhld94** [comment]
> Thanks! Can you add a test for it?

I will add test. In addition, I found this bug also happens for other method, e.g. HashMethodFixedString,... The following query will crash the server. 

```sql
select 'aaa' from numbers(10) intersect select 'aaa' from numbers(10);
```

Seem a critical bug.

**Comment by canhld94** [comment]
> Also worth fixing all HashMethod*, not only HashMethodOneNumber, for example such query leads to segfault:
> 
> ```
> select '1' from numbers(10) intersect select '1' from numbers(10)
> ```
> 
> If you have time, you can fix it in this PR, if not, I will create a separate PR.

I will added the fix for other types.

**Comment by canhld94** [comment]
> Also worth fixing all HashMethod*, not only HashMethodOneNumber, for example such query leads to segfault:
> 
> ```
> select '1' from numbers(10) intersect select '1' from numbers(10)
> ```
> 
> If you have time, you can fix it in this PR, if not, I will create a separate PR.

I added the fix fo HashMethodString and HashMethodFixedString as well as some tests. IMO only these 3 HashMethod* are affected by this bug. If others method also need adding, I think someone else should add it as I'm not quite familiar with their implementation.

**Review by Avogar — APPROVED** [review]

**Comment by canhld94** [comment]
[The query no. 6 of test 'distinct_in_order' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.distinct_in_order.6)
[The query no. 7 of test 'distinct_in_order' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.distinct_in_order.7)
[The query no. 8 of test 'distinct_in_order' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.distinct_in_order.8)
[The query no. 0 of test 'distinct_in_order' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.distinct_in_order.0)
[The query no. 3 of test 'distinct_in_order' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.distinct_in_order.3)
[The query no. 4 of test 'distinct_in_order' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.distinct_in_order.4)
[The query no. 0 of test 'set' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.set.0)
[The query no. 1 of test 'set' has slowed down](https://s3.amazonaws.com/clickhouse-test-reports/40020/20855838764a4aa429b09961b4e73b176e8af7d6/performance_comparison_[2/4]/report.html#changes-in-performance.set.1)

Looks like checking const every time getting key is not a good idea. I will rewrite the implementation.

File: src/Common/ColumnsHashing.h [hunk 1]
```diff
@@ -39,16 +42,36 @@ struct HashMethodOneNumber
...
         vec = column->getRawData().data;
+        if (isColumnConst(*column))
+        {
+            const_value = unalignedLoad<FieldType>(vec);
+            get_key_holder_impl = [this](size_t /*row*/) { return const_value; };
+        }
+        else
+        {
+            get_key_holder_impl = [this](size_t row) { return unalignedLoad<FieldType>(vec + row * sizeof(FieldType)); };
+        }
```
Comment (Avogar) [inline src/Common/ColumnsHashing.h:74]: Let's remove code duplication from these two constructors to separate method

**Comment by alexey-milovidov** [comment]
@Avogar @canhld94 The Aggregation methods don't expect to work const columns and its implementation is correct.
The caller should materialize const columns before using them.

**Comment by alexey-milovidov** [comment]
The tests should check both INTERSECT and EXCEPT.

**Review by alexey-milovidov — DISMISSED** [review]
.

**Comment by canhld94** [comment]
> @Avogar @canhld94 The Aggregation methods don't expect to work const columns and its implementation is correct. It is reasonable to keep code compact. The caller should materialize const columns before using them.

@alexey-milovidov done materializing column in IntersectOrExceptTransform and add more test. In addition, as your suggestion, I remove some code in HashMethodString to handle const column (added in #37738 to fix server crashing while intersect const column, this PR should fix that as well).

**Review by Avogar — COMMENTED** [review]
Now it looks much better

**Review by Avogar — APPROVED** [review]

**Comment by Avogar** [comment]
Let's wait for the rest of checks and merge it

File: tests/queries/0_stateless/02381_intersect_except_const_column.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+-- Test: intersect return incorrect result for const column
+SELECT 1 FROM numbers(10) INTERSECT SELECT 1 FROM numbers(10);
+SELECT toString(1) FROM numbers(10) INTERSECT SELECT toString(1) FROM numbers(10);
+SELECT '1' FROM numbers(10) INTERSECT SELECT '1' FROM numbers(10);
+
+-- Test: except return incorrect result for const column
+SELECT 2 FROM numbers(10) EXCEPT SELECT 1 FROM numbers(5);
+SELECT toString(2) FROM numbers(10) EXCEPT SELECT toString(1) FROM numbers(5);
+SELECT '2' FROM numbers(10) EXCEPT SELECT '1' FROM numbers(5);
```
[the remark this answers is not in the document]
Comment (canhld94) [reply tests/queries/0_stateless/02381_intersect_except_const_column.sql:13]: I think the CI only run for code file not test file 😂

[the remark this answers is not in the document]
Comment (amosbird) [reply tests/queries/0_stateless/02381_intersect_except_const_column.sql:13]: OK. There is a well-known adage: "all text files should end with a newline". POSIX also has a definition of [what is a line](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206) . I think it's better to apply this rule to all text files.

**Pull request merged**